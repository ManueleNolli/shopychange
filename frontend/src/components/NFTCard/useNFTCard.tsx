import { useCallback, useEffect, useState } from 'react'
import { Placeholder } from '../../assets/AssetsManager'
import { useUserContext } from '../../context/userContext'
import useFetching from '../../hooks/useFetching'
import { Sale } from '../../types/components/Sale'
import { NFTCardProps } from '../../types/props/NFTCardProps'
import {
  handlerQueryErrorCatch,
  saleQuery,
} from '../../services/BackendService'
import CheckImageSrc from '../../utils/CheckImageSrc/CheckImageSrc'

export default function useNFTCard({ nft, saleButton }: NFTCardProps) {
  const { userAddress, blockchainNetworkId } = useUserContext()
  const { isFetching, setIsFetching, isError, setIsError } = useFetching()

  const [image, setImage] = useState<string>(Placeholder)
  const [sale, setSale] = useState<Sale | null>(null)

  const fetchSale = async () => {
    if (!userAddress || !blockchainNetworkId) return

    const response = await handlerQueryErrorCatch(
      saleQuery(blockchainNetworkId, nft.contractAddress, nft.tokenId)
    )

    if (response.error) {
      setIsError(true)
      setIsFetching(false)
      return
    }

    const sale = response.result.sale as Sale
    setSale(sale)
  }

  const updateSale = useCallback(async () => {
    fetchSale()
  }, [])

  useEffect(() => {
    const fetchImage = async () => {
      const exists = await CheckImageSrc(nft.image)
      if (exists) setImage(nft.image)
    }

    fetchImage()

    if (saleButton) fetchSale().then(() => setIsFetching(false))
  }, [])

  return {
    isFetching,
    isError,
    image,
    sale,
    updateSale,
  }
}
