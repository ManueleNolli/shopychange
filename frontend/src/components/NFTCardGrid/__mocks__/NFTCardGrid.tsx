import { divWithoutChildrenMock } from '../../../../mocksUtils'

const NFTCardGridMock = () => divWithoutChildrenMock('NFTCardGrid')
export default NFTCardGridMock
