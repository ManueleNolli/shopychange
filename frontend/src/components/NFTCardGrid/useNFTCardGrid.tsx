import { useNavigate } from 'react-router-dom'
import { NFT } from '../../types/components/NFT'

export default function useNFTCardGrid() {
  const navigate = useNavigate()

  const onNavigation = (nft: NFT) => {
    navigate(`/${nft.contractAddress}/${nft.tokenId}`)
  }

  return {
    onNavigation,
  }
}
