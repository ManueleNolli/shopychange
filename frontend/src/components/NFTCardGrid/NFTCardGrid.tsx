import React from 'react'

import { SimpleGrid } from '@chakra-ui/react'
import { NFTCardGridProps } from '../../types/props/NFTCardGridProps'
import NFTCard from '../NFTCard/NFTCard'
import useNFTCardGrid from './useNFTCardGrid'

export default function NFTCardGrid({
  nfts,
  nftChildren,
  nftChildrenWithProps,
  saleButton,
}: NFTCardGridProps) {
  const { onNavigation } = useNFTCardGrid()

  return (
    <SimpleGrid
      pl="5"
      pr="5"
      spacing="5"
      maxW={{ sm: '600px', md: '2500px' }}
      columns={{ sm: 2, md: 3, lg: 4, xl: 5 }}
    >
      {nfts.map((nft) => (
        <NFTCard
          saleButton={saleButton}
          nft={nft}
          key={`${nft.contractAddress}.${nft.tokenId}`}
          onClick={() => onNavigation(nft)}
        >
          {nftChildren && nftChildren}
          {nftChildrenWithProps && nftChildrenWithProps(nft)}
        </NFTCard>
      ))}
    </SimpleGrid>
  )
}
