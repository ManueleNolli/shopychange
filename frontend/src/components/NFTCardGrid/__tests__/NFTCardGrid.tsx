import React from 'react'
import NFTCardGrid from '../NFTCardGrid'

import { render, fireEvent } from '@testing-library/react'
import useNFTCardGrid from '../useNFTCardGrid'
import useNFTCard from '../../NFTCard/useNFTCard'

const mockData = [
  {
    contractAddress: '0x0000000000',
    tokenId: 0,
    image:
      'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q/0.png',
    name: 'Test NFT',
  },
  {
    contractAddress: '0x0000000000',
    tokenId: 1,
    image:
      'https://ipfs.io/ipfs/bafybeietntip4sq3xmuwiz3s43yb3klrpa3okgtg5deysmbd5hbizhrb2q/1.png',
    name: 'Test NFT 1',
  },
]

jest.mock('../useNFTCardGrid', () => jest.fn())
jest.mock('../../Sell/SellOrIsForSale/SellOrIsForSale')
jest.mock('../../NFTCard/useNFTCard', () => jest.fn())

describe('NFTCardGrid', () => {
  test('renders NFTCardGrid component', () => {
    ;(useNFTCardGrid as jest.Mock).mockReturnValue({
      onNavigation: jest.fn(),
    })
    ;(useNFTCard as jest.Mock).mockReturnValue({
      isFetching: false,
      isError: false,
      image: null,
      sale: null,
      updateSale: jest.fn(),
    })

    const { container } = render(<NFTCardGrid nfts={mockData} />)

    expect(container).toMatchSnapshot()
  })

  test('onClick call onNavigation', () => {
    const mockOnNavigation = jest.fn()
    ;(useNFTCardGrid as jest.Mock).mockReturnValue({
      onNavigation: mockOnNavigation,
    })
    ;(useNFTCard as jest.Mock).mockReturnValue({
      isFetching: false,
      isError: false,
      image: null,
      sale: null,
      updateSale: jest.fn(),
    })

    const { getByTestId } = render(<NFTCardGrid nfts={mockData} />)

    fireEvent.click(getByTestId('onclick-button-0x0000000000-0'))

    expect(mockOnNavigation).toHaveBeenCalled()
  })
})
