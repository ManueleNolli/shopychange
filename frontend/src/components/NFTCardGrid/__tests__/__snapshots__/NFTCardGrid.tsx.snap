// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFTCardGrid renders NFTCardGrid component 1`] = `
<div>
  <div
    class="css-xbmy3v"
  >
    <div
      class="css-18n9l0z"
    >
      <div
        class="css-1oxm8tl"
        role="group"
      >
        <div
          class="css-ub82xt"
        >
          <div
            class="css-rbycdq"
            data-testid="onclick-button-0x0000000000-0"
          >
            <img
              class="chakra-image css-gbtqvj"
            />
          </div>
        </div>
        <div
          class="css-13cp05h"
        >
          <h2
            class="chakra-heading css-gdw5dm"
          >
            Test NFT
          </h2>
        </div>
      </div>
    </div>
    <div
      class="css-18n9l0z"
    >
      <div
        class="css-1oxm8tl"
        role="group"
      >
        <div
          class="css-ub82xt"
        >
          <div
            class="css-rbycdq"
            data-testid="onclick-button-0x0000000000-1"
          >
            <img
              class="chakra-image css-gbtqvj"
            />
          </div>
        </div>
        <div
          class="css-13cp05h"
        >
          <h2
            class="chakra-heading css-gdw5dm"
          >
            Test NFT 1
          </h2>
        </div>
      </div>
    </div>
  </div>
</div>
`;
